File: mtimmerm/bicom101/src/bialib/simplemodel.cpp
//===========================================================================
//  Copyright (C) 1999 Matt Timmermans
//  Free for non-commercial purposes as long as this notice remains intact.
//  To discuss commercial use, mail me at matt@timmermans.org
//===========================================================================

#include "simplemodel.h"

//===========================================================================
//  ProbHeapModelBase: a base class for order0 models that use a probability
//    heap
//===========================================================================

ProbHeapModelBase::ProbHeapModelBase()
{
    int i;
    symzeroindex = 256; // number of symbols
    // probability heap is twice this big
    probheap = new unsigned[symzeroindex << 1];
    for (i = symzeroindex << 1; i--;)
        probheap[i] = 0;
    // initial probabilities
    for (i = 256; i--;)
        AddP(i, 1);
};

ProbHeapModelBase::~ProbHeapModelBase()
{
    delete[] probheap;
}

void ProbHeapModelBase::Encode(BYTE sym, ArithmeticEncoder *dest)
{
    int symbol = sym;
    int i, bit = symzeroindex;
    unsigned low = 0;

    for (i = 1; i < symzeroindex;)
    {
        bit >>= 1;
        i += i;
        if (symbol & bit)
            low += probheap[i++];
    }
    dest->Encode(probheap[1], low, low + probheap[i]);
    Update(symbol);
}

void ProbHeapModelBase::Encode(BYTE sym, ArithmeticEncoder *dest, ExclusionSet &ex)
{
    int symbol = sym;
    int i, bit = symzeroindex;
    unsigned low = 0;
    unsigned expl, exph;
    const BYTE *exs, *exe;

    for (i = 1; i < symzeroindex;)
    {
        bit >>= 1;
        i += i;
        if (symbol & bit)
            low += probheap[i++];
    }
    for (
        exs = ex.Sort(), exe = exs + ex.NumExcls(), expl = exph = 0;
        exs != exe;
        ++exs)
    {
        if (*exs > sym)
            exph += SymP(*exs);
        else
            expl += SymP(*exs);
    }
    low -= expl;
    dest->Encode(probheap[1] - (expl + exph), low, low + probheap[i]);
}

BYTE ProbHeapModelBase::Decode(ArithmeticDecoder *src)
{
    int i;
    BYTE ret;
    unsigned low = 0;
    unsigned p = src->GetP(probheap[1]);

    for (i = 1; i < symzeroindex;)
    {
        i += i;
        if ((p - low) >= probheap[i])
            low += probheap[i++];
    }
    ret = (BYTE)(i - symzeroindex);
    src->Narrow(probheap[1], low, low + probheap[i]);
    Update(ret);
    return (ret);
}

BYTE ProbHeapModelBase::Decode(ArithmeticDecoder *src, ExclusionSet &ex)
{
    int i;
    BYTE ret;
    unsigned low = 0, tp;
    unsigned p1 = probheap[1], p;
    const BYTE *exs = ex.Sort();
    const BYTE *exe = exs + ex.NumExcls();

    while (exe != exs)
    {
        --exe;
        p1 -= SymP(*exe);
    }
    exe = exs + ex.NumExcls();
    p = src->GetP(p1);

    for (i = 1; i < symzeroindex;)
    {
        i += i;
        if ((p - low) >= probheap[i])
            low += probheap[i++];
    }
    p -= low;
    while (
        (exs != exe) &&
        (*exs <= (i - symzeroindex)))
    {
        tp = SymP(*exs);
        p += tp;
        low -= tp;
        ++exs;
        while (p >= (tp = probheap[i]))
        {
            p -= tp;
            low += tp;
            ++i;
        }
    }
    ret = (BYTE)(i - symzeroindex);
    src->Narrow(p1, low, low + probheap[i]);
    return (ret);
}

void ProbHeapModelBase::AddP(int sym, unsigned n)
{
    for (sym += symzeroindex; sym; sym >>= 1)
        probheap[sym] += n;
}

void ProbHeapModelBase::SubP(int sym, unsigned n)
{
    for (sym += symzeroindex; sym; sym >>= 1)
        probheap[sym] -= n;
}

//===========================================================================
//  SimpleAdaptiveModel : An adaptive order0 model that only remembers the
//    last 4096 symbols.  More recent symbols are given higher weights
//===========================================================================

SimpleAdaptiveModel::SimpleAdaptiveModel()
{
    int i;
    symzeroindex = 256; // number of symbols
    // probability heap is twice this big
    probheap = new unsigned[symzeroindex << 1];
    for (i = symzeroindex << 1; i--;)
        probheap[i] = 0;
    // initial probabilities
    for (i = 256; i--;)
        AddP(i, 1);

    // set up window
    for (i = 4096; i--;)
        window[i] = -1; // no symbol
    w0 = window;
    w1 = w0 + 1024;
    w2 = w0 + 2048;
    w3 = w0 + 3072;
}

void SimpleAdaptiveModel::Reset()
{
    int *w, *lim;
    lim = window + 4095;

    for (w = w0; w != w1; w = (w == lim ? window : w + 1))
    {
        if (*w < 0)
            goto DONE;
        SubP(*w, 6);
        *w = -1;
    }
    for (w = w1; w != w2; w = (w == lim ? window : w + 1))
    {
        if (*w < 0)
            goto DONE;
        SubP(*w, 4);
        *w = -1;
    }
    for (w = w2; w != w3; w = (w == lim ? window : w + 1))
    {
        if (*w < 0)
            goto DONE;
        SubP(*w, 3);
        *w = -1;
    }
    for (w = w3; w != w0; w = (w == lim ? window : w + 1))
    {
        if (*w < 0)
            goto DONE;
        SubP(*w, 2);
        *w = -1;
    }

DONE:
    return;
}

void SimpleAdaptiveModel::Update(int symbol)
{
    w1 = ((w1 == window) ? w1 + 4095 : w1 - 1);
    if (*w1 >= 0) // 6
        SubP(*w1, 2);
    w2 = ((w2 == window) ? w2 + 4095 : w2 - 1);
    if (*w2 >= 0) // 4
        SubP(*w2, 1);
    w3 = ((w3 == window) ? w3 + 4095 : w3 - 1);
    if (*w3 >= 0) // 3
        SubP(*w3, 1);
    w0 = ((w0 == window) ? w0 + 4095 : w0 - 1);
    if (*w0 >= 0) // 2
        SubP(*w0, 2);
    *w0 = symbol;
    AddP(symbol, 6);
}

//===========================================================================
//  SimpleDecayModel : An adaptive order0 model that decays old counts
//    gradually
//===========================================================================

SimpleDecayModel::SimpleDecayModel(unsigned inicounts, unsigned decayshift)
{
    m_decayshift = decayshift;
    int i;
    for (i = symzeroindex + 255; i >= symzeroindex; --i)
    {
        probheap[i] = inicounts;
    }
    for (; i; --i)
    {
        probheap[i] = probheap[i << 1] + probheap[(i << 1) + 1];
    }
}

void SimpleDecayModel::Update(int symbol)
{
    AddP(symbol, (P1() >> m_decayshift) + 1);
    if (P1() > MAXP1)
    {
        int i;
        for (i = symzeroindex + 255; i >= symzeroindex; --i)
        {
            probheap[i] = (probheap[i] + 1) >> 1;
        }
        for (; i; --i)
        {
            probheap[i] = probheap[i << 1] + probheap[(i << 1) + 1];
        }
    }
}

File: mtimmerm/biacode/src/python/anthropic/simple_model.py
class SimpleAdaptiveModel:
    def __init__(self, num_symbols):
        # Round num_symbols to next power of 2
        self.sym_zero_index = 1
        while self.sym_zero_index < num_symbols:
            self.sym_zero_index *= 2

        # Probability heap is twice the size
        self.prob_heap = [0] * (self.sym_zero_index * 2)
        self.prob1 = 0

        # Initial probabilities
        for i in range(num_symbols):
            self.add_p(i, 1)

        # Circular window for tracking recent symbols
        self.window = [-1] * 4096
        self.indices = [1024, 2048, 3072, 0]

    def update(self, symbol):
        # Circular buffer navigation and probability updates
        weights = [2, 1, 1, 2]

        for i in range(len(self.indices)):
            self.indices[i] = (self.indices[i] - 1) % 4096
            if self.window[self.indices[i]] >= 0:
                self.sub_p(self.window[self.indices[i]], weights[i])

        self.window[self.indices[-1]] = symbol
        self.add_p(symbol, 6)

    def reset(self):
        # Reset probabilities and window
        weights = [6, 4, 3, 2]

        for i in range(len(self.indices)):
            j = (i + len(self.indices) - 1) % len(self.indices)
            curr = self.indices[j]
            while curr != self.indices[i]:
                if self.window[curr] < 0:
                    break
                self.sub_p(self.window[curr], weights[i])
                self.window[curr] = -1
                curr = (curr + 1) % 4096

    def get_sym_range(self, symbol):
        # Get probability range for a symbol
        low = 0
        bit = self.sym_zero_index
        i = 1

        while i < self.sym_zero_index:
            bit //= 2
            i *= 2
            if symbol & bit:
                low += self.prob_heap[i]
                i += 1

        return low, low + self.prob_heap[i]

    def get_symbol(self, p):
        # Get symbol for a given probability
        low = 0
        i = 1

        while i < self.sym_zero_index:
            i *= 2
            if (p - low) >= self.prob_heap[i]:
                low += self.prob_heap[i]
                i += 1

        return i - self.sym_zero_index, low, low + self.prob_heap[i]

    def add_p(self, sym, n):
        # Add probability for a symbol
        sym += self.sym_zero_index
        while sym:
            self.prob_heap[sym] += n
            sym //= 2
        self.prob1 = self.prob_heap[1]

    def sub_p(self, sym, n):
        # Subtract probability for a symbol
        sym += self.sym_zero_index
        while sym:
            self.prob_heap[sym] -= n
            sym //= 2
        self.prob1 = self.prob_heap[1]

    def prob_one(self):
        return self.prob1
